src/Nat.mo

```diff
@@ -1,72 +1,299 @@
-/// Natural numbers
+/// Utility functions for working with natural numbers.
 ///
 /// Most operations on natural numbers (e.g. addition) are available as built-in operators (e.g. `1 + 1`).
 /// This module provides equivalent functions and `Text` conversion.
+///
+/// Import from the base library to use this module.
+/// ```motoko name=import
+/// import Nat "mo:base/Nat";
+/// ```
 
 import Int "Int";
 import Order "Order";
 import Prim "mo:⛔";
+import Char "Char";
 
 module {
 
   /// Infinite precision natural numbers.
   public type Nat = Prim.Types.Nat;
 
-  /// Conversion.
-  public let toText : Nat -> Text = Int.toText;
+  /// Converts a natural number to its textual representation.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// Nat.toText 1234 // => "1234"
+  /// ```
+  public func toText(n : Nat) : Text = Int.toText n;
+
+  /// Creates a natural number from its textual representation. Returns `null`
+  /// if the input is not a valid natural number.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// Nat.fromText "1234" // => ?1234
+  /// ```
+  public func fromText(text : Text) : ?Nat {
+    if (text == "") {
+      return null
+    };
+    var n = 0;
+    for (c in text.chars()) {
+      if (Char.isDigit(c)) {
+        let charAsNat = Prim.nat32ToNat(Prim.charToNat32(c) -% Prim.charToNat32('0'));
+        n := n * 10 + charAsNat
+      } else {
+        return null
+      }
+    };
+    ?n
+  };
 
   /// Returns the minimum of `x` and `y`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// Nat.min(1, 2) // => 1
+  /// ```
   public func min(x : Nat, y : Nat) : Nat {
     if (x < y) { x } else { y }
   };
 
   /// Returns the maximum of `x` and `y`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// Nat.max(1, 2) // => 2
+  /// ```
   public func max(x : Nat, y : Nat) : Nat {
     if (x < y) { y } else { x }
   };
 
-  /// Returns `x == y`.
+  /// Equality function for Nat types.
+  /// This is equivalent to `x == y`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// ignore Nat.equal(1, 1); // => true
+  /// 1 == 1 // => true
+  /// ```
+  ///
+  /// Note: The reason why this function is defined in this library (in addition
+  /// to the existing `==` operator) is so that you can use it as a function
+  /// value to pass to a higher order function. It is not possible to use `==`
+  /// as a function value at the moment.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// import Buffer "mo:base/Buffer";
+  ///
+  /// let buffer1 = Buffer.Buffer<Nat>(3);
+  /// let buffer2 = Buffer.Buffer<Nat>(3);
+  /// Buffer.equal(buffer1, buffer2, Nat.equal) // => true
+  /// ```
```

Comment:
And nice example demonstrating where it makes really sense to use the equal function.